src/compiler/abstract_compiler.nit

```diff
@@ -2689,10 +2722,15 @@ redef class AMethPropdef
 		var pname = mpropdef.mproperty.name
 		var cname = mpropdef.mclassdef.mclass.name
 		var ret = mpropdef.msignature.return_mtype
-		if ret != null then
+                var compiler = v.compiler
+                # WARNING: we must not resolve the return type when it's a functional type.
+                # Otherwise, we get a compile error exactly here. This weird behavior doesn't affect
+                # the inner mecanics of callref since the return type is already solved by
+                # `routine_ref_call`
+                if ret != null and not compiler.all_routine_types_name.has(cname) then
 			ret = v.resolve_for(ret, arguments.first)
 		end
-		if pname != "==" and pname != "!=" then
+		if pname != "==" and pname != "!=" and pname != "call" and not compiler.all_routine_types_name.has(cname) then
```

Comment:
Why this is necessary? (add comment and also add a test case in `/tests`)

Reply:
I added comments, but there's no test case for this. The problem is in the if where a call to adapt_signature is made. The intern `call` function has to do some processing over its argument before generating the callsite.

Reply:
Please add a test case. This is weird and I could not reproduce it.